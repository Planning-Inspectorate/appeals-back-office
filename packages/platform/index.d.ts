import { HandlerFunction } from 'got';

declare module '@pins/platform' {
	export function createTtlHandler(): HandlerFunction;
	export function validatePostcode(value: string): boolean;
	export function validatePastDate(value: Date | string | number): boolean;
	export function validateFutureDate(value: Date | string | number): boolean;
	export function yesterday(): Date;
}

export interface LocalPlanningDepartment {
	FID: number;
	LPA21CD: string;
	LPA21NM: string;
	Co_terminous: 'Y' | 'N';
}
