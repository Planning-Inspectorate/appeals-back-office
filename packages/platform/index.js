export * from './http/ttl.js';
export * from './util/date.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
