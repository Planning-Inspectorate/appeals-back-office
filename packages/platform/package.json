{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=22.0.0 <= 22.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "test": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci",
    "test:cov": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci --coverage",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.3",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.4",
    "@types/lodash-es": "^4.17.12",
    "cross-env": "^7.0.3",
    "jest": "^29.7.0",
    "type-fest": "^4.37.0"
  }
}
