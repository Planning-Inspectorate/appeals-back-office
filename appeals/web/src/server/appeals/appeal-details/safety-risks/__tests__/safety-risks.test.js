import { parseHtml } from '@pins/platform';
import supertest from 'supertest';
import { appealData } from '#testing/app/fixtures/referencedata.js';
import { createTestEnvironment } from '#testing/index.js';
import nock from 'nock';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);
const baseUrl = '/appeals-service/appeal-details';

describe('safety-risks', () => {
	beforeEach(installMockApi);
	afterEach(teardown);

	describe('GET /change/:source', () => {
		it('should render changeSafetyRisksAccess page for the LPA answer when source is LPA from appeals details', async () => {
			const appealId = appealData.appealId;
			const response = await request.get(`${baseUrl}/${appealId}/safety-risks/change/lpa`);

			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('Are there any potential safety risks?</h1>');

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain(
				'name="safetyRisksRadio" type="radio" value="yes"'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'<textarea class="govuk-textarea" id="safety-risks-details" name="safetyRisksDetails" rows="3">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="safetyRisksRadio" type="radio" value="no"'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});

		it('should render changeSafetyRisksAccess page for the appellant answer when source is appellant from appeals details', async () => {
			const appealId = appealData.appealId;
			const response = await request.get(`${baseUrl}/${appealId}/safety-risks/change/appellant`);

			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('Are there any potential safety risks?</h1>');

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain(
				'name="safetyRisksRadio" type="radio" value="yes"'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'<textarea class="govuk-textarea" id="safety-risks-details" name="safetyRisksDetails" rows="3">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="safetyRisksRadio" type="radio" value="no"'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});

		it('should render changeSafetyRisksAccess page for the LPA answer when source is LPA from LPA questionnaire', async () => {
			const appealId = appealData.appealId;
			const lpaQuestionnaireId = appealData.lpaQuestionnaireId;
			const response = await request.get(
				`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/safety-risks/change/lpa`
			);

			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('Are there any potential safety risks?</h1>');

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain(
				'name="safetyRisksRadio" type="radio" value="yes"'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'<textarea class="govuk-textarea" id="safety-risks-details" name="safetyRisksDetails" rows="3">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="safetyRisksRadio" type="radio" value="no"'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});

		it('should render changeSafetyRisksAccess page for the appellant answer when source is appellant from appellant case', async () => {
			const appealId = appealData.appealId;

			const response = await request.get(
				`${baseUrl}/${appealId}/appellant-case/safety-risks/change/appellant`
			);

			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('Are there any potential safety risks?</h1>');

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain(
				'name="safetyRisksRadio" type="radio" value="yes"'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'<textarea class="govuk-textarea" id="safety-risks-details" name="safetyRisksDetails" rows="3">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="safetyRisksRadio" type="radio" value="no"'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});
	});

	describe('POST /change/:source', () => {
		it('should re-render changeSafetyRisksAccess page with an error when isRequired is "yes" but details is empty', async () => {
			const appealId = appealData.appealId.toString();

			const invalidData = {
				safetyRisksRadio: 'yes',
				safetyRisksDetails: ''
			};
			const response = await request
				.post(`${baseUrl}/${appealId}/safety-risks/change/lpa`)
				.send(invalidData);

			expect(response.statusCode).toBe(200);

			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('Are there any potential safety risks?</h1>');

			const errorSummaryHtml = parseHtml(response.text, {
				rootElement: '.govuk-error-summary',
				skipPrettyPrint: true
			}).innerHTML;

			expect(errorSummaryHtml).toContain('There is a problem</h2>');
			expect(errorSummaryHtml).toContain('Enter health and safety risks</a>');
		});

		it('should re-render changeSafetyRisksAccess page with an error when safetyRisksDetails exceeds 1000 characters', async () => {
			const appealId = appealData.appealId.toString();
			const invalidData = {
				safetyRisksRadio: 'yes',
				safetyRisksDetails: 'a'.repeat(1001) // Creates string of 1001 'a' characters
			};

			const response = await request
				.post(`${baseUrl}/${appealId}/safety-risks/change/lpa`)
				.send(invalidData);

			expect(response.statusCode).toBe(200);

			const elementInnerHtml = parseHtml(response.text).innerHTML;
			const errorSummaryHtml = parseHtml(response.text, {
				rootElement: '.govuk-error-summary',
				skipPrettyPrint: true
			}).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(errorSummaryHtml).toContain('There is a problem</h2>');
			expect(errorSummaryHtml).toContain(
				'Health and safety risks must be 1000 characters or less</a>'
			);
		});

		it('should re-direct to appeals-details page when data is valid for LPA and came from appeal-details', async () => {
			const appealId = appealData.appealId;
			const lpaQuestionnaireId = appealData.lpaQuestionnaireId;
			const validData = {
				safetyRisksRadio: 'yes',
				safetyRisksDetails: 'Details'
			};

			nock('http://test/')
				.patch(`/appeals/${appealId}/lpa-questionnaires/${lpaQuestionnaireId}`)
				.reply(200, {
					...validData
				});

			const response = await request
				.post(`${baseUrl}/${appealId}/safety-risks/change/lpa`)
				.send(validData);

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe('Found. Redirecting to /appeals-service/appeal-details/1');
		});

		it('should re-direct to appeals-details page when data is valid for appellant and came from appeal-details', async () => {
			const appealId = appealData.appealId;
			const appellantCaseId = appealData.appellantCaseId;
			const validData = {
				safetyRisksRadio: 'yes',
				safetyRisksDetails: 'Details'
			};

			nock('http://test/')
				.patch(`/appeals/${appealId}/appellant-cases/${appellantCaseId}`)
				.reply(200, {
					...validData
				});

			const response = await request
				.post(`${baseUrl}/${appealId}/safety-risks/change/appellant`)
				.send(validData);

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe('Found. Redirecting to /appeals-service/appeal-details/1');
		});

		it('should re-direct to appellant case page when data is valid for appellant and came from appellant-case', async () => {
			const appealId = appealData.appealId;
			const appellantCaseId = appealData.appellantCaseId;
			const validData = {
				safetyRisksRadio: 'yes',
				safetyRisksDetails: 'Details'
			};

			nock('http://test/')
				.patch(`/appeals/${appealId}/appellant-cases/${appellantCaseId}`)
				.reply(200, {
					...validData
				});

			const response = await request
				.post(`${baseUrl}/${appealId}/appellant-case/safety-risks/change/appellant`)
				.send(validData);

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				'Found. Redirecting to /appeals-service/appeal-details/1/appellant-case'
			);
		});

		it('should re-direct to lpaq page when data is valid for LPA and came from lpa-questionnaire', async () => {
			const appealId = appealData.appealId;
			const lpaQuestionnaireId = appealData.lpaQuestionnaireId;
			const validData = {
				safetyRisksRadio: 'yes',
				safetyRisksDetails: 'Details'
			};

			nock('http://test/')
				.patch(`/appeals/${appealId}/lpa-questionnaires/${lpaQuestionnaireId}`)
				.reply(200, {
					...validData
				});

			const response = await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/safety-risks/change/lpa`
				)
				.send(validData);

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				`Found. Redirecting to /appeals-service/appeal-details/1/lpa-questionnaire/${lpaQuestionnaireId}`
			);
		});
	});
});
