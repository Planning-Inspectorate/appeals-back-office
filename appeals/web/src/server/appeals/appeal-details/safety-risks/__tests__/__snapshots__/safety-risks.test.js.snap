// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`safety-risks GET /change/:source should render changeSafetyRisksAccess page for the LPA answer when source is LPA from LPA questionnaire 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change the site health and safety risks (LPA answer)</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="safetyRisksRadio" name="safetyRisksRadio"
                            type="radio" value="yes" data-aria-controls="conditional-safetyRisksRadio">
                            <label class="govuk-label govuk-radios__label" for="safetyRisksRadio">Yes</label>
                        </div>
                        <div class="govuk-radios__conditional govuk-radios__conditional--hidden"
                        id="conditional-safetyRisksRadio">
                            <div class="govuk-form-group">
                                <label class="govuk-label" for="safety-risk-details">Health and safety risks (LPA details)</label>
                                <textarea class="govuk-textarea"
                                id="safety-risk-details" name="safetyRisksDetails" rows="3"></textarea>
                            </div>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="safetyRisksRadio-2" name="safetyRisksRadio"
                            type="radio" value="no" checked>
                            <label class="govuk-label govuk-radios__label" for="safetyRisksRadio-2">No</label>
                        </div>
                    </div>
                </div>
                <button type="submit" class="govuk-button" data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`safety-risks GET /change/:source should render changeSafetyRisksAccess page for the LPA answer when source is LPA from appeals details 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change the site health and safety risks (LPA answer)</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="safetyRisksRadio" name="safetyRisksRadio"
                            type="radio" value="yes" data-aria-controls="conditional-safetyRisksRadio">
                            <label class="govuk-label govuk-radios__label" for="safetyRisksRadio">Yes</label>
                        </div>
                        <div class="govuk-radios__conditional govuk-radios__conditional--hidden"
                        id="conditional-safetyRisksRadio">
                            <div class="govuk-form-group">
                                <label class="govuk-label" for="safety-risk-details">Health and safety risks (LPA details)</label>
                                <textarea class="govuk-textarea"
                                id="safety-risk-details" name="safetyRisksDetails" rows="3"></textarea>
                            </div>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="safetyRisksRadio-2" name="safetyRisksRadio"
                            type="radio" value="no" checked>
                            <label class="govuk-label govuk-radios__label" for="safetyRisksRadio-2">No</label>
                        </div>
                    </div>
                </div>
                <button type="submit" class="govuk-button" data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`safety-risks GET /change/:source should render changeSafetyRisksAccess page for the appellant answer when source is appellant from appeals details 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change the site health and safety risks (appellant answer)</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="safetyRisksRadio" name="safetyRisksRadio"
                            type="radio" value="yes" checked data-aria-controls="conditional-safetyRisksRadio">
                            <label class="govuk-label govuk-radios__label" for="safetyRisksRadio">Yes</label>
                        </div>
                        <div class="govuk-radios__conditional" id="conditional-safetyRisksRadio">
                            <div class="govuk-form-group">
                                <label class="govuk-label" for="safety-risk-details">Health and safety risks (appellant details)</label>
                                <textarea class="govuk-textarea"
                                id="safety-risk-details" name="safetyRisksDetails" rows="3">Dogs on site</textarea>
                            </div>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="safetyRisksRadio-2" name="safetyRisksRadio"
                            type="radio" value="no">
                            <label class="govuk-label govuk-radios__label" for="safetyRisksRadio-2">No</label>
                        </div>
                    </div>
                </div>
                <button type="submit" class="govuk-button" data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`safety-risks GET /change/:source should render changeSafetyRisksAccess page for the appellant answer when source is appellant from appellant case 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change the site health and safety risks (appellant answer)</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="safetyRisksRadio" name="safetyRisksRadio"
                            type="radio" value="yes" checked data-aria-controls="conditional-safetyRisksRadio">
                            <label class="govuk-label govuk-radios__label" for="safetyRisksRadio">Yes</label>
                        </div>
                        <div class="govuk-radios__conditional" id="conditional-safetyRisksRadio">
                            <div class="govuk-form-group">
                                <label class="govuk-label" for="safety-risk-details">Health and safety risks (appellant details)</label>
                                <textarea class="govuk-textarea"
                                id="safety-risk-details" name="safetyRisksDetails" rows="3">Dogs on site</textarea>
                            </div>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="safetyRisksRadio-2" name="safetyRisksRadio"
                            type="radio" value="no">
                            <label class="govuk-label govuk-radios__label" for="safetyRisksRadio-2">No</label>
                        </div>
                    </div>
                </div>
                <button type="submit" class="govuk-button" data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`safety-risks POST /change/:source should re-render changeSafetyRisksAccess page with an error when isRequired is "yes" but details is empty 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-error-summary" data-module="govuk-error-summary">
                <div role="alert">
                    <h2 class="govuk-error-summary__title"> There is a problem</h2>
                    <div class="govuk-error-summary__body">
                        <ul class="govuk-list govuk-error-summary__list">
                            <li><a href="#safetyRisksDetails">Provide details of health and safety risks</a>
                            </li>
                        </ul>
                    </div>
                </div>
            </div>
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change the site health and safety risks (LPA answer)</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="safetyRisksRadio" name="safetyRisksRadio"
                            type="radio" value="yes" checked data-aria-controls="conditional-safetyRisksRadio">
                            <label class="govuk-label govuk-radios__label" for="safetyRisksRadio">Yes</label>
                        </div>
                        <div class="govuk-radios__conditional" id="conditional-safetyRisksRadio">
                            <div class="govuk-form-group">
                                <label class="govuk-label" for="safety-risk-details">Health and safety risks (LPA details)</label>
                                <textarea class="govuk-textarea"
                                id="safety-risk-details" name="safetyRisksDetails" rows="3"></textarea>
                            </div>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="safetyRisksRadio-2" name="safetyRisksRadio"
                            type="radio" value="no">
                            <label class="govuk-label govuk-radios__label" for="safetyRisksRadio-2">No</label>
                        </div>
                    </div>
                </div>
                <button type="submit" class="govuk-button" data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;
